includes/utils/class-amp-dom-document.php

```diff
@@ -0,0 +1,313 @@
+			'<head>',
+			'<head><meta http-equiv="content-type" content="text/html; charset=' . self::AMP_ENCODING . '">',
+			$source
+		);
+
+		$success = parent::loadHTML( $source, $options );
+
+		if ( $success ) {
+			$this->encoding = self::AMP_ENCODING;
+			$head           = $this->getElementsByTagName( 'head' )->item( 0 );
+			$head->removeChild( $head->firstChild );
+		}
+
+		return $success;
+	}
+
+	/**
+	 * Dumps the internal document into a string using HTML formatting.
+	 *
+	 * @link  https://php.net/manual/domdocument.savehtml.php
+	 *
+	 * @param DOMNode $node Optional. Parameter to output a subset of the document.
+	 * @return string The HTML, or false if an error occurred.
+	 */
+	public function saveHTML( DOMNode $node = null ) {
+		// Force-add http-equiv charset to make DOMDocument behave as it should.
+		// See: http://php.net/manual/en/domdocument.loadhtml.php#78243.
+		$head    = $this->getElementsByTagName( 'head' )->item( 0 );
+		$charset = AMP_DOM_Utils::create_node(
+			$this,
+			'meta',
+			[
+				'http-equiv' => 'content-type',
+				'content'    => 'text/html; charset=' . self::AMP_ENCODING,
+			]
+		);
+		$head->insertBefore( $charset, $head->firstChild );
+
+		return str_replace( '<meta http-equiv="content-type" content="text/html; charset=' . self::AMP_ENCODING . '">', '', parent::saveHTML( $node ) );
+	}
+
+	/**
+	 * Maybe add the <head> and/or <body> tag(s).
+	 *
+	 * @param string $content Content to add the head or body tags to.
+	 * @return string Adapted content.
+	 */
+	private function maybe_add_head_or_body( $content ) {
+		$substring = substr( $content, 0, 5000 );
+		if ( false === strpos( $substring, '<body' ) ) {
+			if ( false === strpos( $substring, '<head>' ) ) {
+				// Create the required HTML structure if none exists yet.
+				$content = "<html><head></head><body>{$content}</body></html>";
+			} else {
+				// <head> seems to be present without <body>.
+				$content = preg_replace( '#</head>(.*)</html>#', '</head><body>$1</body>', $content );
+			}
+		} elseif ( false === strpos( $substring, '<head>' ) ) {
+			// Create a <head> element if none exists yet.
+			$content = str_replace( '<body', '<head></head><body', $content );
+		}
+
+		return $content;
+	}
+
+	/**
+	 * Adapt the encoding of the content.
+	 *
+	 * @param string $source Source content to adapt the encoding of.
+	 * @return string Adapted content.
+	 */
+	private function adapt_encoding( $source ) {
+		// No encoding was provided, so we need to guess.
+		if ( self::UNKNOWN_ENCODING === $this->original_encoding ) {
+			$this->original_encoding = mb_detect_encoding( $source );
+		}
+
+		// Guessing the encoding seems to have failed, so we assume UTF-8 instead.
+		if ( empty( $this->original_encoding ) ) {
+			$this->original_encoding = self::AMP_ENCODING;
+		}
+
+		$this->original_encoding = $this->sanitize_encoding( $this->original_encoding );
+
+		$target = mb_convert_encoding( $source, self::AMP_ENCODING, $this->original_encoding );
```

Comment:
This will need to short-circuit if the `mbstring` extension is not loaded, since we do not currently include it among the `$_amp_required_extensions`.

Reply:
Actually, I suggest that if a document is not `utf-8` and we cannot convert into `utf-8` that we throw an exception. AMP would just not be available.